v2/netconf/server/ssh/server_test.go

```diff
@@ -0,0 +1,159 @@
+
+	sshcfg, err := PasswordConfig(TestUserName, TestPassword)
+	assert.NoError(t, err)
+
+	ctx := WithSshTrace(context.Background(), DefaultLoggingHooks)
+	server, err := NewServer(ctx, "localhost", 0, sshcfg, handlerFactory())
+	assert.NotNil(t, server)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	//----------------------------
+
+	sshConfig := &xssh.ClientConfig{
+		User:            TestUserName,
+		Auth:            []xssh.AuthMethod{xssh.Password(TestPassword)},
+		HostKeyCallback: xssh.InsecureIgnoreHostKey(),
+	}
+
+	ctx = context.Background()
+	tr, err := client.NewSSHTransport(ctx, sshConfig, fmt.Sprintf("localhost:%d", server.Port()), "netconf")
+	assert.NoError(t, err, "Not expecting new transport to fail")
+	defer tr.Close()
+
+	tr.Write([]byte("hello"))
+	buffer := make([]byte, 7)
+	tr.Read(buffer)
+	assert.Equal(t, ">hello<", string(buffer))
+}
+
+func TestServerListenFailure(t *testing.T) {
+
+	sshcfg, err := PasswordConfig(TestUserName, TestPassword)
+	assert.NoError(t, err)
+
+	ctx := WithSshTrace(context.Background(), DefaultLoggingHooks)
+	server, err := NewServer(ctx, "9.9.9.9", 9999, sshcfg, handlerFactory())
+	assert.Nil(t, server)
+	assert.Error(t, err)
+}
+
+func TestServerConnectionFailure(t *testing.T) {
+
+	sshcfg, err := PasswordConfig(TestUserName, TestPassword)
+	assert.NoError(t, err)
+
+	ctx := WithSshTrace(context.Background(), DefaultLoggingHooks)
+	server, err := NewServer(ctx, "localhost", 0, sshcfg, handlerFactory())
+	assert.NotNil(t, server)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	//----------------------------
+
+	sshConfig := &xssh.ClientConfig{
+		User:            TestUserName,
+		Auth:            []xssh.AuthMethod{xssh.Password("WrongPassword")},
+		HostKeyCallback: xssh.InsecureIgnoreHostKey(),
+	}
+
+	ctx = context.Background()
+	_, err = client.NewSSHTransport(ctx, sshConfig, fmt.Sprintf("localhost:%d", server.Port()), "netconf")
+	assert.Error(t, err, "Not expecting new transport to succeed")
+	assert.Contains(t, err.Error(), "authenticate")
+}
+
+func TestServerDiagnosticTraceHooks(t *testing.T) {
+
+	sshcfg, err := PasswordConfig(TestUserName, TestPassword)
+	assert.NoError(t, err)
+
+	ctx := WithSshTrace(context.Background(), DiagnosticLoggingHooks)
+	server, err := NewServer(ctx, "localhost", 0, sshcfg, handlerFactory())
+	assert.NotNil(t, server)
+	assert.NoError(t, err)
+	defer server.Close()
+
+	//----------------------------
+
+	sshConfig := &xssh.ClientConfig{
+		User:            TestUserName,
+		Auth:            []xssh.AuthMethod{xssh.Password(TestPassword)},
+		HostKeyCallback: xssh.InsecureIgnoreHostKey(),
+	}
+
+	ctx = context.Background()
+	tr, err := client.NewSSHTransport(ctx, sshConfig, fmt.Sprintf("localhost:%d", server.Port()), "netconf")
+	assert.NoError(t, err, "Not expecting new transport to fail")
+	defer tr.Close()
+
+	tr.Write([]byte("hello"))
+	buffer := make([]byte, 7)
+	tr.Read(buffer)
+	assert.Equal(t, ">hello<", string(buffer))
```

Comment:
can you clarify how this tests Diag Trace Hooks, not obvious without running code why pattern matching on hello does this.

Reply:
In truth it doesn't - the test is only there for coverage.   
Could parameterise the default diagnostic trace hooks and inject a mock/test function but it seems to complicate what is otherwise very simple code.